### PR TITLE
WIP: Add erf and erfc

### PR DESCRIPTION
**Pull request by SebastianM-C**
Implement `erfc` as discussed in #113
- In IntervalArithmetic.jl:
- [ ] Implement `erf(x::Interval)` and `erfc(x::Interval)` by using `erf(y::BigFloat)` and `erfc(y::BigFloat)` with monotonicity and directed rounding.
- [ ] Implement the inverse functions `erfinv` and `erfcinv` using interval Newton method.
- In IntervalContractors.jl:
- [ ] Follow the pattern that is defined there to add the reverse function `erf_rev`.

### CODE REVIEW

**Comment by dpsanders** [comment]
Thanks for the contribution!

At the moment, you're missing `erf(x, r)` when `x` is a `Float64`.

**Comment by dpsanders** [comment]
I was also discussing with @lbenet that since this introduces a dependency on SpecialFunctions.jl, we probably should split this out into a new IntervalSpecialFunctions.jl package.

**Comment by SebastianM-C** [comment]
Should I continue with the PR here or it would be better to see what can be done regarding the split?

**Comment by dpsanders** [comment]
Feel free to continue here, then we can just move the branch to a new repo when it's ready.

File: src/intervals/functions.jl [hunk 1]
```diff
@@ -266,3 +266,22 @@ for f in (:log, :log2, :log10, :log1p)
...
+end
+
+function erf(a::Interval{T}) where T
+    isempty(a) && return a
+    @round( erf(a.lo), erf(a.hi) )
+end
+
+function erfc(a::Interval{T}) where T
+    isempty(a) && return a
+    @round( erf(a.hi), erfc(a.lo) )
```
Comment (dpsanders) [inline src/intervals/functions.jl]: `erfc(a.hi)`

**Comment by dpsanders** [comment]
Will need some tests. May as well put them in a new test file.

File: src/intervals/functions.jl [hunk 2]
```diff
@@ -266,3 +266,22 @@ for f in (:log, :log2, :log10, :log1p)
...
         end
 end
+
+for f in (:erf, :erfc)
+    @eval function($f)(x::BigFloat, r::RoundingMode)
+        setrounding(BigFloat, r) do
+            ($f)(x)
+        end
+    end
+
```
Comment (SebastianM-C) [inline src/intervals/functions.jl]: For the `Float64` case is `@eval ($f)(a::Interval{Float64}) = atomic(Interval{Float64}, $f(big53(a)))` ok? (I saw this for the `exp2` function, but it is not clear to me what `big53` is)

**Comment by SebastianM-C** [comment]
Should I add the `SpecialFunctions` package to REQUIRE so that the CI can work?

File: src/intervals/functions.jl (hunk 2 above)
Comment (dpsanders) [reply src/intervals/functions.jl]: `big53` is defined in the package. It converts the interval to a BigFloat interval with 53 bits of precision (i.e. the same as Interval{Float64}). You don't need `atomic`, just `convert`.

**Comment by dpsanders** [comment]
Yes, go ahead and add `SpecialFunctions` to REQUIRE for CI.

File: src/intervals/functions.jl [hunk 3]
```diff
@@ -266,3 +266,24 @@ for f in (:log, :log2, :log10, :log1p)
...
+    @eval ($f)(a::Interval{Float64}) = ($f)(big53(a))
```
Comment (dpsanders) [inline src/intervals/functions.jl]: You need to `convert` the result back to an `Interval{Float64}`.

**Comment by SebastianM-C** [comment]
I think that the first part is almost complete. I will add some tests, but I am not sure how to build them (or what should be tested).

**Comment by dpsanders** [comment]
You can just choose some random intervals and write literal tests like
```
@test erf(1..2) == Interval(0.8427007929497148, 0.9953222650189528)
```
or test that certain numbers are in the result of `erf(1..2)` etc.

**Comment by dpsanders** [comment]
@SebastianM-C Once you add a couple more tests, I think we can merge this and do the rest in separate PRs.

**Comment by SebastianM-C** [comment]
Should the checklist at the top be moved in #113 ?

Also: do you think that the tests are ok? I don't understand why the absolute coverage decreased.

**Comment by dpsanders** [comment]
The tests look good, thanks!
Not sure about the coverage, wouldn't worry about it.

I would close that issue since it's now solved and move the checklist to a new issue.

**Comment by dpsanders** [comment]
Thanks a lot!

**Comment by dpsanders** [comment]
Now that I think about it, if we use the interval Newton method for the inverse functions, that will need IntervalRootFinding (or a custom implementation). In that case we should definitely split this out into a separate package.

**Comment by dpsanders** [comment]
Wow, it really is slow:
```
x = 1..2
julia> @btime erf($x)
  61.184 μs (63 allocations: 2.41 KiB)
Interval(0.8427007929497148, 0.9953222650189528)

julia> @btime sin($x)
  119.898 ns (0 allocations: 0 bytes)
Interval(0.8414709848078965, 1.0)
```

**Comment by SebastianM-C** [comment]
Glad I could help! Thanks for guiding me 😄

**Pull request merged**